[PyTorch] Add backend aot_eager_decomp_partition_with_mode

silverlakeli opened the pull request:
Summary:
## Why
To make it possible to run torch dispatch mode inside compiled modules. This is to enable running MemoryTrackerMode (in next diff) to collect memory usage of compiled modules.

## What
Add a backend aot_eager_decomp_partition_with_mode.
Add an enable_log to the backend to control the compilation logging (which can be very verbose and slow the run of mode)

Test Plan:
unittest

E2e tested in the next diff which shows the memory read from the mode passed to this backend is very close to the actual job's memory snapshot.

Differential Revision: D67227144




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

basilwong commented on torch/_functorch/partitioners.py:
```diff
@@ -1850,10 +1850,10 @@ def classify_nodes(joint_module):
         # Log total theoretical activations stored
         total_activations_size_gb = sum(_size_of(i) for i in saved_values) / 1e9
-        log.debug("Theoretical Activations Stored: %.2f GB", total_activations_size_gb)
+        log.info("Theoretical Activations Stored: %.2f GB", total_activations_size_gb)
```
Why change to info instead of debug?

eellison commented:
@silverlakeli does this replace the need for https://github.com/pytorch/pytorch/issues/136446 ? or what's the additional value of inductor here - cc @exclamaforte

silverlakeli commented:
> @silverlakeli does this replace the need for #136446 ? or what's the additional value of inductor here - cc @exclamaforte

Yes it achieves similar effect as the #136446 and no need for #136446 for now. Thanks.

silverlakeli committed:
[PyTorch] Add backend aot_eager_decomp_partition_with_mode (#143250)

Summary:

## Why
To make it possible to run torch dispatch mode inside compiled modules. This is to enable running MemoryTrackerMode (in next diff) to collect memory usage of compiled modules.

## What
Add a backend aot_eager_decomp_partition_with_mode.
Add an enable_log to the backend to control the compilation logging (which can be very verbose and slow the run of mode)

Test Plan:
unittest

E2e tested in the next diff which shows the memory read from the mode passed to this backend is very close to the actual job's memory snapshot.

Created/updated Pull Request #143250

Reviewed By: bdhirsh

Differential Revision: D67227144

izaitsevfb commented:
@pytorchbot merge -f "flaky job"
